test/lib/runtime/broker_test.exs

```diff
@@ -1238,4 +1281,72 @@ defmodule Ask.Runtime.BrokerTest do
     end)
   end
 
+  defp init_completed_dispositions_shouldnt_be_rejected_tests() do
+    insert_channel = fn mode ->
+      insert(:channel, settings: TestChannel.new() |> TestChannel.settings(), type: mode)
+    end
+
+    ivr_channel = insert_channel.("ivr")
+    sms_channel = insert_channel.("sms")
+
+    quiz = insert(:questionnaire, steps: @completed_dispositions_shouldnt_be_rejected_dummy_steps)
+
+    insert_survey = fn mode ->
+      insert(
+        :survey,
+        schedule: Schedule.always(),
+        state: "running",
+        questionnaires: [quiz],
+        mode: [[mode]],
+        count_partial_results: true,
+        quota_vars: ["City"]
+      )
+    end
+
+    ivr_survey = insert_survey.("ivr")
+    sms_survey = insert_survey.("sms")
+
+    insert_bucket = fn survey ->
+      insert(:quota_bucket, condition: %{City: "GBA"}, quota: 1, count: 0, survey: survey)
+    end
+
+    insert_bucket.(ivr_survey)
+    insert_bucket.(sms_survey)
+
+    insert_group = fn survey ->
+      insert(:respondent_group, survey: survey, respondents_count: 1) |> Repo.preload([:channels])
+    end
+
+    ivr_group = insert_group.(ivr_survey)
+    sms_group = insert_group.(sms_survey)
+
+    insert_group_channel = fn mode, channel, group ->
+      RespondentGroupChannel.changeset(%RespondentGroupChannel{}, %{
+        respondent_group_id: group.id,
+        channel_id: channel.id,
+        mode: mode
+      })
+      |> Repo.insert()
+    end
+
+    insert_group_channel.("ivr", ivr_channel, ivr_group)
+    insert_group_channel.("sms", sms_channel, sms_group)
+
+    %{
+      ivr_respondent: insert(:respondent, survey: ivr_survey, respondent_group: ivr_group),
+      sms_respondent: insert(:respondent, survey: sms_survey, respondent_group: sms_group)
+    }
+  end
+
+  defp broker_poll(), do: Broker.handle_info(:poll, nil)
```

Comment:
you can directly use `Broker.poll` that is intended for tests 🙂 
```elixir
  # Makes the broker performs a poll on the surveys.
  # This method is intended to be used by tests.
  def poll do
    GenServer.call(@server_ref, :poll)
  end
```

Reply:
I've tried it without success. The error I get is the following:

```
root@75554624e88a:/app# mix test test/lib/runtime/broker_test.exs:204
Including tags: [line: "204"]
Excluding tags: [:test]

  1) test #1752 - respondents in completed dispositions shouldn't be rejected doesn't reject interim partial - IVR (Ask.Runtime.BrokerTest)
     test/lib/runtime/broker_test.exs:204
     ** (exit) exited in: GenServer.call({:global, Ask.Runtime.Broker}, :poll, 5000)
         ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
     stacktrace:
       (elixir) lib/gen_server.ex:766: GenServer.call/3
       test/lib/runtime/broker_test.exs:195: Ask.Runtime.BrokerTest.__ex_unit_setup_5/1
       test/lib/runtime/broker_test.exs:1: Ask.Runtime.BrokerTest.__ex_unit__/2

Finished in 2.2 seconds
50 tests, 1 failure, 49 skipped
```

`Broker.poll()` doesn't appear to be the same as `Broker.handle_info(:poll, nil)` Do you have a clue about their differences?

Reply:
By the way, I also copied `broker_poll` from [here](https://github.com/instedd/surveda/blob/1873d9f8988e0a5f97e99a0f1bfe07c91944819c/test/lib/runtime/retries_histogram_test.exs#L453) in the retries histogram tests. If it stays, I should also avoid code duplication here, as we did with the `respondent_reply` function.

Reply:
Right, I had forgot about this 🤦‍♀️ 

If you take a look at the trace, the process - GenServer - isn't alive, so the `GenServer.call` explotes since it doesn't find a process with that name. For using directly `Broker.poll` you have to first have the process running as a GenServer. You can take a look at `test/lib/runtime/survey_test.exs` where this is done: first they start it with 
```elixir 
{:ok, broker} = Broker.start_link
Broker.poll
```

Using directly the `Broker.handle_info(:poll, nil)` is the way of "by-passing" the GenServer and just using the logic it uses but without being run on a separate process - For what I understand, we can use it like this because the `Broker` does nothing with its internal state

Reply:
👍 I should learn deeply about how Surveda manages its processes. Maybe, in one of our pairing sessions, we can take a look together 😉